// @letta-ai/letta-client/resources/archives/archives.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Archives = void 0;
const tslib_1 = require("../../internal/tslib.js");
const resource_1 = require("../../core/resource.js");
const PassagesAPI = tslib_1.__importStar(require("./passages.js"));
const passages_1 = require("./passages.js");
const pagination_1 = require("../../core/pagination.js");
const headers_1 = require("../../internal/headers.js");
const path_1 = require("../../internal/utils/path.js");
class Archives extends resource_1.APIResource {
    constructor() {
        super(...arguments);
        this.passages = new PassagesAPI.Passages(this._client);
    }
    /**
     * Create a new archive.
     */
    create(body, options) {
        return this._client.post('/v1/archives/', { body, ...options });
    }
    /**
     * Get a single archive by its ID.
     */
    retrieve(archiveID, options) {
        return this._client.get((0, path_1.path) `/v1/archives/${archiveID}`, options);
    }
    /**
     * Update an existing archive's name and/or description.
     */
    update(archiveID, body, options) {
        return this._client.patch((0, path_1.path) `/v1/archives/${archiveID}`, { body, ...options });
    }
    /**
     * Get a list of all archives for the current organization with optional filters
     * and pagination.
     */
    list(query = {}, options) {
        return this._client.getAPIList('/v1/archives/', (pagination_1.ArrayPage), { query, ...options });
    }
    /**
     * Delete an archive by its ID.
     */
    delete(archiveID, options) {
        return this._client.delete((0, path_1.path) `/v1/archives/${archiveID}`, {
            ...options,
            headers: (0, headers_1.buildHeaders)([{ Accept: '*/*' }, options?.headers]),
        });
    }
}
exports.Archives = Archives;
Archives.Passages = passages_1.Passages;
