// @letta-ai/letta-client/resources/mcp-servers/mcp-servers.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.McpServers = void 0;
const tslib_1 = require("../../internal/tslib.js");
const resource_1 = require("../../core/resource.js");
const ToolsAPI = tslib_1.__importStar(require("./tools.js"));
const tools_1 = require("./tools.js");
const headers_1 = require("../../internal/headers.js");
const path_1 = require("../../internal/utils/path.js");
class McpServers extends resource_1.APIResource {
    constructor() {
        super(...arguments);
        this.tools = new ToolsAPI.Tools(this._client);
    }
    /**
     * Add a new MCP server to the Letta MCP server config
     */
    create(body, options) {
        return this._client.post('/v1/mcp-servers/', { body, ...options });
    }
    /**
     * Get a specific MCP server
     */
    retrieve(mcpServerID, options) {
        return this._client.get((0, path_1.path) `/v1/mcp-servers/${mcpServerID}`, options);
    }
    /**
     * Update an existing MCP server configuration
     */
    update(mcpServerID, body, options) {
        return this._client.patch((0, path_1.path) `/v1/mcp-servers/${mcpServerID}`, { body, ...options });
    }
    /**
     * Get a list of all configured MCP servers
     */
    list(options) {
        return this._client.get('/v1/mcp-servers/', options);
    }
    /**
     * Delete an MCP server by its ID
     */
    delete(mcpServerID, options) {
        return this._client.delete((0, path_1.path) `/v1/mcp-servers/${mcpServerID}`, {
            ...options,
            headers: (0, headers_1.buildHeaders)([{ Accept: '*/*' }, options?.headers]),
        });
    }
    /**
     * Refresh tools for an MCP server by:
     *
     * 1. Fetching current tools from the MCP server
     * 2. Deleting tools that no longer exist on the server
     * 3. Updating schemas for existing tools
     * 4. Adding new tools from the server
     *
     * Returns a summary of changes made.
     */
    refresh(mcpServerID, params = {}, options) {
        const { agent_id } = params ?? {};
        return this._client.patch((0, path_1.path) `/v1/mcp-servers/${mcpServerID}/refresh`, {
            query: { agent_id },
            ...options,
        });
    }
}
exports.McpServers = McpServers;
McpServers.Tools = tools_1.Tools;
