// @letta-ai/letta-client/resources/tools.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Tools = void 0;
const resource_1 = require("../core/resource.js");
const pagination_1 = require("../core/pagination.js");
const path_1 = require("../internal/utils/path.js");
class Tools extends resource_1.APIResource {
    /**
     * Create a new tool
     */
    create(body, options) {
        return this._client.post('/v1/tools/', { body, ...options });
    }
    /**
     * Get a tool by ID
     */
    retrieve(toolID, options) {
        return this._client.get((0, path_1.path) `/v1/tools/${toolID}`, options);
    }
    /**
     * Update an existing tool
     */
    update(toolID, body, options) {
        return this._client.patch((0, path_1.path) `/v1/tools/${toolID}`, { body, ...options });
    }
    /**
     * Get a list of all tools available to agents.
     */
    list(query = {}, options) {
        return this._client.getAPIList('/v1/tools/', (pagination_1.ArrayPage), { query, ...options });
    }
    /**
     * Delete a tool by name
     */
    delete(toolID, options) {
        return this._client.delete((0, path_1.path) `/v1/tools/${toolID}`, options);
    }
    /**
     * Search tools using semantic search.
     *
     * Requires tool embedding to be enabled (embed_tools=True). Uses vector search,
     * full-text search, or hybrid mode to find tools matching the query.
     *
     * Returns tools ranked by relevance with their search scores.
     */
    search(body, options) {
        return this._client.post('/v1/tools/search', { body, ...options });
    }
    /**
     * Create or update a tool
     */
    upsert(body, options) {
        return this._client.put('/v1/tools/', { body, ...options });
    }
}
exports.Tools = Tools;
