// @letta-ai/letta-client/resources/runs/runs.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Runs = void 0;
const tslib_1 = require("../../internal/tslib.js");
const resource_1 = require("../../core/resource.js");
const RunsMessagesAPI = tslib_1.__importStar(require("./messages.js"));
const messages_1 = require("./messages.js");
const StepsAPI = tslib_1.__importStar(require("./steps.js"));
const steps_1 = require("./steps.js");
const TraceAPI = tslib_1.__importStar(require("./trace.js"));
const trace_1 = require("./trace.js");
const UsageAPI = tslib_1.__importStar(require("./usage.js"));
const usage_1 = require("./usage.js");
const pagination_1 = require("../../core/pagination.js");
const path_1 = require("../../internal/utils/path.js");
class Runs extends resource_1.APIResource {
    constructor() {
        super(...arguments);
        this.messages = new RunsMessagesAPI.Messages(this._client);
        this.usage = new UsageAPI.Usage(this._client);
        this.steps = new StepsAPI.Steps(this._client);
        this.trace = new TraceAPI.Trace(this._client);
    }
    /**
     * Get the status of a run.
     */
    retrieve(runID, options) {
        return this._client.get((0, path_1.path) `/v1/runs/${runID}`, options);
    }
    /**
     * List all runs.
     */
    list(query = {}, options) {
        return this._client.getAPIList('/v1/runs/', (pagination_1.ArrayPage), { query, ...options });
    }
}
exports.Runs = Runs;
Runs.Messages = messages_1.Messages;
Runs.Usage = usage_1.Usage;
Runs.Steps = steps_1.Steps;
Runs.Trace = trace_1.Trace;
