// @letta-ai/letta-client/resources/templates/templates.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Templates = void 0;
const tslib_1 = require("../../internal/tslib.js");
const resource_1 = require("../../core/resource.js");
const AgentsAPI = tslib_1.__importStar(require("./agents.js"));
const agents_1 = require("./agents.js");
const path_1 = require("../../internal/utils/path.js");
class Templates extends resource_1.APIResource {
    constructor() {
        super(...arguments);
        this.agents = new AgentsAPI.Agents(this._client);
    }
    /**
     * Creates a new template from an existing agent or agent file
     */
    create(body, options) {
        return this._client.post('/v1/templates', { body, ...options });
    }
    /**
     * Updates the current working version of a template from an agent file
     */
    update(templateName, body, options) {
        return this._client.patch((0, path_1.path) `/v1/templates/${templateName}`, { body, ...options });
    }
    /**
     * Deletes all versions of a template with the specified name
     */
    delete(templateName, body, options) {
        return this._client.delete((0, path_1.path) `/v1/templates/${templateName}`, { body, ...options });
    }
    /**
     * Rollback the current working version of a template to a previous saved version.
     * If the current version has unsaved changes, they will be automatically saved as
     * a new version before rollback.
     */
    rollback(templateName, body, options) {
        return this._client.post((0, path_1.path) `/v1/templates/${templateName}/rollback`, { body, ...options });
    }
    /**
     * Saves the current version of the template as a new version
     */
    save(templateName, body = {}, options) {
        return this._client.post((0, path_1.path) `/v1/templates/${templateName}/save`, { body, ...options });
    }
}
exports.Templates = Templates;
Templates.Agents = agents_1.Agents;
