// @letta-ai/letta-code-sdk/dist/index.js
import { createRequire } from "node:module";
var __create = Object.create;
var __getProtoOf = Object.getPrototypeOf;
var __defProp = Object.defineProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __toESM = (mod, isNodeMode, target) => {
  target = mod != null ? __create(__getProtoOf(mod)) : {};
  const to = isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target;
  for (let key of __getOwnPropNames(mod))
    if (!__hasOwnProp.call(to, key))
      __defProp(to, key, {
        get: () => mod[key],
        enumerable: true
      });
  return to;
};
var __require = /* @__PURE__ */ createRequire(import.meta.url);

// src/transport.ts
import { spawn } from "node:child_process";
import { createInterface } from "node:readline";
function sdkLog(tag, ...args) {
  if (process.env.DEBUG_SDK)
    console.error(`[SDK-Transport] [${tag}]`, ...args);
}
function buildCliArgs(options) {
  const args = [
    "--output-format",
    "stream-json",
    "--input-format",
    "stream-json"
  ];
  if (options.conversationId) {
    args.push("--conversation", options.conversationId);
  } else if (options.agentId) {
    args.push("--agent", options.agentId);
    if (options.newConversation) {
      args.push("--new");
    } else if (options.defaultConversation) {
      args.push("--conversation", "default");
    }
  } else if (options.createOnly) {
    args.push("--new-agent");
  } else if (options.newConversation) {
    args.push("--new");
  }
  if (options.model) {
    args.push("-m", options.model);
  }
  if (options.includePartialMessages) {
    args.push("--include-partial-messages");
  }
  if (options.embedding) {
    args.push("--embedding", options.embedding);
  }
  if (options.systemPrompt !== undefined) {
    if (typeof options.systemPrompt === "string") {
      const validPresets = [
        "default",
        "letta-claude",
        "letta-codex",
        "letta-gemini",
        "claude",
        "codex",
        "gemini"
      ];
      if (validPresets.includes(options.systemPrompt)) {
        args.push("--system", options.systemPrompt);
      } else {
        args.push("--system-custom", options.systemPrompt);
      }
    } else {
      args.push("--system", options.systemPrompt.preset);
      if (options.systemPrompt.append) {
        args.push("--system-append", options.systemPrompt.append);
      }
    }
  }
  if (options.memory !== undefined && !options.agentId) {
    if (options.memory.length === 0) {
      args.push("--init-blocks", "");
    } else {
      const presetNames = [];
      const memoryBlocksJson = [];
      for (const item of options.memory) {
        if (typeof item === "string") {
          presetNames.push(item);
        } else if ("blockId" in item) {
          memoryBlocksJson.push(item);
        } else {
          memoryBlocksJson.push(item);
        }
      }
      if (memoryBlocksJson.length > 0) {
        args.push("--memory-blocks", JSON.stringify(memoryBlocksJson));
        if (presetNames.length > 0) {
          console.warn("[letta-code-sdk] Using custom memory blocks. " + `Preset blocks are ignored when custom blocks are provided: ${presetNames.join(", ")}`);
        }
      } else if (presetNames.length > 0) {
        args.push("--init-blocks", presetNames.join(","));
      }
    }
  }
  if (!options.agentId) {
    if (options.persona !== undefined) {
      args.push("--block-value", `persona=${options.persona}`);
    }
    if (options.human !== undefined) {
      args.push("--block-value", `human=${options.human}`);
    }
  }
  if (options.permissionMode === "bypassPermissions") {
    args.push("--yolo");
  } else if (options.permissionMode && options.permissionMode !== "default") {
    args.push("--permission-mode", options.permissionMode);
  }
  if (options.allowedTools) {
    args.push("--allowedTools", options.allowedTools.join(","));
  }
  if (options.disallowedTools) {
    args.push("--disallowedTools", options.disallowedTools.join(","));
  }
  if (options.tags && options.tags.length > 0) {
    args.push("--tags", options.tags.join(","));
  }
  if (options.memfs === true) {
    args.push("--memfs");
  } else if (options.memfs === false) {
    args.push("--no-memfs");
  }
  if (options.memfsStartup !== undefined) {
    args.push("--memfs-startup", options.memfsStartup);
  }
  if (options.skillSources !== undefined) {
    const sources = [...new Set(options.skillSources)];
    if (sources.length === 0) {
      args.push("--no-skills");
    } else {
      args.push("--skill-sources", sources.join(","));
    }
  }
  if (options.systemInfoReminder === false) {
    args.push("--no-system-info-reminder");
  }
  if (options.sleeptime?.trigger !== undefined) {
    args.push("--reflection-trigger", options.sleeptime.trigger);
  }
  if (options.sleeptime?.behavior !== undefined) {
    args.push("--reflection-behavior", options.sleeptime.behavior);
  }
  if (options.sleeptime?.stepCount !== undefined) {
    args.push("--reflection-step-count", String(options.sleeptime.stepCount));
  }
  return args;
}

class SubprocessTransport {
  options;
  process = null;
  stdout = null;
  messageQueue = [];
  messageResolvers = [];
  closed = false;
  agentId;
  wireMessageCount = 0;
  lastMessageAt = 0;
  stderrLines = [];
  constructor(options = {}) {
    this.options = options;
  }
  async connect() {
    const args = this.buildArgs();
    const cliPath = await this.findCli();
    sdkLog("connect", `CLI: ${cliPath}`);
    sdkLog("connect", `args: ${args.join(" ")}`);
    sdkLog("connect", `cwd: ${this.options.cwd || process.cwd()}`);
    sdkLog("connect", `permissionMode: ${this.options.permissionMode || "default"}`);
    this.process = spawn("node", [cliPath, ...args], {
      cwd: this.options.cwd || process.cwd(),
      stdio: ["pipe", "pipe", "pipe"],
      env: { ...process.env }
    });
    const pid = this.process.pid;
    sdkLog("connect", `CLI process spawned, pid=${pid}`);
    if (!this.process.stdout || !this.process.stdin) {
      throw new Error("Failed to create subprocess pipes");
    }
    this.stdout = createInterface({
      input: this.process.stdout,
      crlfDelay: Infinity
    });
    this.stdout.on("line", (line) => {
      if (!line.trim())
        return;
      try {
        const msg = JSON.parse(line);
        this.handleMessage(msg);
      } catch {
        sdkLog("stdout", `[non-JSON] ${line.slice(0, 500)}`);
      }
    });
    if (this.process.stderr) {
      this.process.stderr.on("data", (data) => {
        const msg = data.toString().trim();
        if (msg) {
          console.error("[letta-code-sdk] CLI stderr:", msg);
          this.stderrLines.push(msg);
        }
      });
    }
    this.process.on("close", (code, signal) => {
      if (code !== 0 && code !== null) {
        console.error(`[letta-code-sdk] CLI process exited with code ${code}`);
      }
      sdkLog("close", `CLI process exited: pid=${pid} code=${code} signal=${signal} wireMessages=${this.wireMessageCount} msSinceLastMsg=${this.lastMessageAt ? Date.now() - this.lastMessageAt : 0} pendingResolvers=${this.messageResolvers.length} queueLen=${this.messageQueue.length}`);
      this.closed = true;
      for (const resolve of this.messageResolvers) {
        resolve(null);
      }
      this.messageResolvers = [];
    });
    this.process.on("error", (err) => {
      console.error("[letta-code-sdk] CLI process error:", err);
      this.closed = true;
    });
  }
  async write(data) {
    if (!this.process?.stdin || this.closed) {
      const err = new Error(`Transport not connected (closed=${this.closed}, pid=${this.process?.pid}, stdin=${!!this.process?.stdin})`);
      sdkLog("write", err.message);
      throw err;
    }
    const payload = data;
    sdkLog("write", `type=${payload.type} subtype=${payload.request?.subtype || payload.response?.subtype || "N/A"}`);
    this.process.stdin.write(JSON.stringify(data) + `
`);
  }
  async read() {
    if (this.messageQueue.length > 0) {
      return this.messageQueue.shift();
    }
    if (this.closed) {
      sdkLog("read", `returning null (closed), total wireMessages=${this.wireMessageCount}`);
      return null;
    }
    sdkLog("read", `waiting for next message (resolvers=${this.messageResolvers.length + 1}, queue=${this.messageQueue.length})`);
    return new Promise((resolve) => {
      this.messageResolvers.push(resolve);
    });
  }
  async* messages() {
    while (true) {
      const msg = await this.read();
      if (msg === null) {
        sdkLog("messages", `iterator ending (closed=${this.closed}, wireMessages=${this.wireMessageCount})`);
        break;
      }
      yield msg;
    }
  }
  close() {
    sdkLog("close", `explicit close called (wireMessages=${this.wireMessageCount}, pendingResolvers=${this.messageResolvers.length}, pid=${this.process?.pid})`);
    if (this.process) {
      this.process.stdin?.end();
      this.process.kill();
      this.process = null;
    }
    this.closed = true;
    for (const resolve of this.messageResolvers) {
      resolve(null);
    }
    this.messageResolvers = [];
  }
  get isClosed() {
    return this.closed;
  }
  getStderr() {
    return this.stderrLines.join(`
`);
  }
  handleMessage(msg) {
    this.wireMessageCount++;
    this.lastMessageAt = Date.now();
    const wirePayload = msg;
    const msgType = wirePayload.message_type || wirePayload.subtype || "";
    sdkLog("wire", `#${this.wireMessageCount} type=${msg.type} ${msgType ? `msg_type=${msgType}` : ""} resolvers=${this.messageResolvers.length} queue=${this.messageQueue.length}`);
    if (msg.type === "result") {
      const result = wirePayload;
      sdkLog("wire", `RESULT: subtype=${result.subtype} stop_reason=${result.stop_reason || "N/A"} duration=${result.duration_ms}ms resultLen=${result.result?.length || 0}`);
    }
    if (msg.type === "system" && "subtype" in msg && msg.subtype === "init") {
      this.agentId = msg.agent_id;
      sdkLog("wire", `INIT: agent_id=${this.agentId}`);
    }
    if (msg.type === "control_request") {
      const req = wirePayload;
      sdkLog("wire", `CONTROL_REQUEST: id=${req.request_id} subtype=${req.request?.subtype} tool=${req.request?.tool_name || "N/A"}`);
    }
    if (this.messageResolvers.length > 0) {
      const resolve = this.messageResolvers.shift();
      resolve(msg);
    } else {
      this.messageQueue.push(msg);
    }
  }
  buildArgs() {
    return buildCliArgs(this.options);
  }
  async findCli() {
    const { existsSync } = await import("node:fs");
    const { dirname, join } = await import("node:path");
    const { fileURLToPath } = await import("node:url");
    if (process.env.LETTA_CLI_PATH && existsSync(process.env.LETTA_CLI_PATH)) {
      return process.env.LETTA_CLI_PATH;
    }
    try {
      const { createRequire: createRequire2 } = await import("node:module");
      const require2 = createRequire2(import.meta.url);
      const resolved = require2.resolve("@letta-ai/letta-code");
      if (existsSync(resolved)) {
        return resolved;
      }
    } catch {}
    const __filename2 = fileURLToPath(import.meta.url);
    const __dirname2 = dirname(__filename2);
    const localPaths = [
      join(__dirname2, "../../@letta-ai/letta-code/letta.js"),
      join(__dirname2, "../../../letta-code-prod/letta.js"),
      join(__dirname2, "../../../letta-code/letta.js")
    ];
    for (const p of localPaths) {
      if (existsSync(p)) {
        return p;
      }
    }
    throw new Error("Letta Code CLI not found. Set LETTA_CLI_PATH or install @letta-ai/letta-code.");
  }
}

// src/interactiveToolPolicy.ts
var INTERACTIVE_APPROVAL_TOOLS = new Set([
  "AskUserQuestion",
  "EnterPlanMode",
  "ExitPlanMode"
]);
var RUNTIME_USER_INPUT_TOOLS = new Set(["AskUserQuestion", "ExitPlanMode"]);
var HEADLESS_AUTO_ALLOW_TOOLS = new Set(["EnterPlanMode"]);
function requiresRuntimeUserInput(toolName) {
  return RUNTIME_USER_INPUT_TOOLS.has(toolName);
}
function isHeadlessAutoAllowTool(toolName) {
  return HEADLESS_AUTO_ALLOW_TOOLS.has(toolName);
}

// src/session.ts
function sessionLog(tag, ...args) {
  if (process.env.DEBUG_SDK)
    console.error(`[SDK-Session] [${tag}]`, ...args);
}
var MAX_BUFFERED_STREAM_MESSAGES = 100;

class Session {
  options;
  transport;
  _agentId = null;
  _sessionId = null;
  _conversationId = null;
  initialized = false;
  externalTools = new Map;
  streamQueue = [];
  streamResolvers = [];
  pumpPromise = null;
  pumpClosed = false;
  droppedStreamMessages = 0;
  sendGeneration = 0;
  lastCompletedRunIds = new Set;
  controlResponseWaiters = new Map;
  constructor(options = {}) {
    this.options = options;
    this.transport = new SubprocessTransport(options);
    if (options.tools) {
      for (const tool of options.tools) {
        this.externalTools.set(tool.name, tool);
      }
    }
  }
  async initialize() {
    if (this.initialized) {
      throw new Error("Session already initialized");
    }
    sessionLog("init", "connecting transport...");
    await this.transport.connect();
    sessionLog("init", "transport connected, sending initialize request");
    await this.transport.write({
      type: "control_request",
      request_id: "init_1",
      request: { subtype: "initialize" }
    });
    sessionLog("init", "waiting for init message from CLI...");
    for await (const msg of this.transport.messages()) {
      sessionLog("init", `received wire message: type=${msg.type}`);
      if (msg.type === "control_request") {
        const handled = await this.handleControlRequest(msg);
        if (!handled) {
          const wireMsgAny = msg;
          sessionLog("init", `DROPPED unsupported control_request: subtype=${wireMsgAny.request?.subtype || "N/A"}`);
        }
        continue;
      }
      if (msg.type === "system" && "subtype" in msg && msg.subtype === "init") {
        const initMsg = msg;
        this._agentId = initMsg.agent_id;
        this._sessionId = initMsg.session_id;
        this._conversationId = initMsg.conversation_id;
        this.initialized = true;
        this.startBackgroundPump();
        if (this.externalTools.size > 0) {
          await this.registerExternalTools();
        }
        const allTools = [
          ...initMsg.tools,
          ...Array.from(this.externalTools.keys())
        ];
        sessionLog("init", `initialized: agent=${initMsg.agent_id} conversation=${initMsg.conversation_id} model=${initMsg.model} tools=${allTools.length} (${this.externalTools.size} external)`);
        return {
          type: "init",
          agentId: initMsg.agent_id,
          sessionId: initMsg.session_id,
          conversationId: initMsg.conversation_id,
          model: initMsg.model,
          tools: allTools,
          memfsEnabled: initMsg.memfs_enabled,
          skillSources: initMsg.skill_sources,
          systemInfoReminderEnabled: initMsg.system_info_reminder_enabled,
          sleeptime: initMsg.reflection_trigger && initMsg.reflection_behavior && typeof initMsg.reflection_step_count === "number" ? {
            trigger: initMsg.reflection_trigger,
            behavior: initMsg.reflection_behavior,
            stepCount: initMsg.reflection_step_count
          } : undefined
        };
      }
    }
    const stderr = this.transport.getStderr();
    const detail = stderr ? `
CLI stderr:
${stderr}` : "";
    sessionLog("init", `ERROR: transport closed before init message received${detail}`);
    throw new Error(`Failed to initialize session - no init message received${detail}`);
  }
  async send(message) {
    if (!this.initialized) {
      sessionLog("send", "auto-initializing (not yet initialized)");
      await this.initialize();
    }
    const preview = typeof message === "string" ? message.slice(0, 100) : Array.isArray(message) ? `[multimodal: ${message.length} parts]` : String(message).slice(0, 100);
    sessionLog("send", `sending message: ${preview}${typeof message === "string" && message.length > 100 ? "..." : ""}`);
    if (this.streamQueue.length > 0) {
      sessionLog("send", `clearing ${this.streamQueue.length} stale messages from previous turn`);
      this.streamQueue.length = 0;
    }
    await this.transport.write({
      type: "user",
      message: { role: "user", content: message }
    });
    this.sendGeneration++;
    sessionLog("send", `message written to transport (generation=${this.sendGeneration})`);
  }
  async* stream() {
    const streamStart = Date.now();
    const minGeneration = this.sendGeneration;
    let yieldCount = 0;
    let staleCount = 0;
    let staleRunIdCount = 0;
    let gotResult = false;
    const currentStreamRunIds = new Set;
    const staleRunIds = new Set(this.lastCompletedRunIds);
    this.startBackgroundPump();
    sessionLog("stream", `starting stream (agent=${this._agentId}, conversation=${this._conversationId}, generation=${minGeneration})`);
    while (true) {
      const bufferedMsg = await this.nextBufferedMessage();
      if (!bufferedMsg) {
        break;
      }
      if (bufferedMsg.generation < minGeneration) {
        staleCount++;
        sessionLog("stream", `discarding stale message: type=${bufferedMsg.message.type} generation=${bufferedMsg.generation} (current=${minGeneration})`);
        continue;
      }
      if (bufferedMsg.runId && staleRunIds.has(bufferedMsg.runId)) {
        staleRunIdCount++;
        sessionLog("stream", `discarding stale message: type=${bufferedMsg.message.type} runId=${bufferedMsg.runId}`);
        continue;
      }
      if (bufferedMsg.runId) {
        currentStreamRunIds.add(bufferedMsg.runId);
      }
      const sdkMsg = bufferedMsg.message;
      yieldCount++;
      sessionLog("stream", `yield #${yieldCount}: type=${sdkMsg.type}${sdkMsg.type === "result" ? ` success=${sdkMsg.success} error=${sdkMsg.error || "none"}` : ""}`);
      yield sdkMsg;
      if (sdkMsg.type === "result") {
        gotResult = true;
        this.updateCompletedRunIds(sdkMsg.runIds, currentStreamRunIds);
        break;
      }
    }
    const elapsed = Date.now() - streamStart;
    sessionLog("stream", `stream ended: duration=${elapsed}ms yielded=${yieldCount} staleFiltered=${staleCount} staleRunIdFiltered=${staleRunIdCount} dropped=${this.droppedStreamMessages} gotResult=${gotResult}`);
    if (!gotResult) {
      sessionLog("stream", "WARNING: stream ended WITHOUT a result message -- transport may have closed unexpectedly");
    }
  }
  startBackgroundPump() {
    if (this.pumpPromise) {
      return;
    }
    this.pumpClosed = false;
    this.pumpPromise = this.runBackgroundPump().catch((err) => {
      sessionLog("pump", `ERROR: ${err instanceof Error ? err.message : String(err)}`);
    }).finally(() => {
      this.pumpClosed = true;
      this.resolveAllStreamWaiters(null);
    });
  }
  async runBackgroundPump() {
    sessionLog("pump", "background pump started");
    const indexToToolCallId = new Map;
    for await (const wireMsg of this.transport.messages()) {
      const wireMsgAny = wireMsg;
      if (wireMsg.type === "control_request") {
        const handled = await this.handleControlRequest(wireMsg);
        if (!handled) {
          sessionLog("pump", `DROPPED unsupported control_request: subtype=${wireMsgAny.request?.subtype || "N/A"}`);
        }
        continue;
      }
      if (wireMsg.type === "control_response") {
        const respMsg = wireMsg;
        const requestId = respMsg.response?.request_id;
        if (requestId && this.controlResponseWaiters.has(requestId)) {
          const resolve = this.controlResponseWaiters.get(requestId);
          this.controlResponseWaiters.delete(requestId);
          resolve(respMsg.response);
        } else {
          sessionLog("pump", `DROPPED unmatched control_response: request_id=${requestId ?? "N/A"}`);
        }
        continue;
      }
      const messageType = wireMsgAny.message_type;
      if (wireMsg.type === "message" && (messageType === "tool_call_message" || messageType === "approval_request_message")) {
        const toolCalls = wireMsgAny.tool_calls;
        const toolCall = wireMsgAny.tool_call;
        const tc = toolCalls?.[0] || toolCall;
        if (tc) {
          const fnObj = tc.function;
          const tcId = tc.tool_call_id ?? tc.id;
          const tcIndex = tc.index;
          if (tcId && tcIndex !== undefined) {
            indexToToolCallId.set(tcIndex, tcId);
          } else if (!tcId && tcIndex !== undefined) {
            const resolvedId = indexToToolCallId.get(tcIndex);
            if (resolvedId) {
              if (fnObj) {
                tc.id = resolvedId;
              } else {
                tc.tool_call_id = resolvedId;
              }
            }
          }
        }
      }
      const sdkMsg = this.transformMessage(wireMsg);
      if (sdkMsg) {
        this.enqueueStreamMessage(sdkMsg);
      } else {
        sessionLog("pump", `DROPPED wire message: type=${wireMsg.type} message_type=${wireMsgAny.message_type || "N/A"} subtype=${wireMsgAny.subtype || "N/A"}`);
      }
    }
    sessionLog("pump", "background pump ended");
  }
  async handleControlRequest(controlReq) {
    const subtype = controlReq.request.subtype;
    sessionLog("pump", `control_request: subtype=${subtype} tool=${controlReq.request.tool_name || "N/A"}`);
    if (subtype === "can_use_tool") {
      await this.handleCanUseTool(controlReq.request_id, controlReq.request);
      return true;
    }
    if (subtype === "execute_external_tool") {
      const rawReq = controlReq.request;
      await this.handleExecuteExternalTool(controlReq.request_id, {
        subtype: "execute_external_tool",
        tool_call_id: rawReq.tool_call_id,
        tool_name: rawReq.tool_name,
        input: rawReq.input
      });
      return true;
    }
    return false;
  }
  enqueueStreamMessage(msg) {
    const bufferedMsg = {
      message: msg,
      generation: this.sendGeneration,
      runId: this.getMessageRunId(msg)
    };
    if (this.streamResolvers.length > 0) {
      const resolve = this.streamResolvers.shift();
      resolve(bufferedMsg);
      return;
    }
    if (this.streamQueue.length >= MAX_BUFFERED_STREAM_MESSAGES) {
      this.streamQueue.shift();
      this.droppedStreamMessages++;
      sessionLog("pump", `stream queue overflow: dropped oldest message (total_dropped=${this.droppedStreamMessages}, max=${MAX_BUFFERED_STREAM_MESSAGES})`);
    }
    this.streamQueue.push(bufferedMsg);
  }
  async nextBufferedMessage() {
    if (this.streamQueue.length > 0) {
      return this.streamQueue.shift();
    }
    if (this.pumpClosed) {
      return null;
    }
    return new Promise((resolve) => {
      this.streamResolvers.push(resolve);
    });
  }
  resolveAllStreamWaiters(msg) {
    for (const resolve of this.streamResolvers) {
      resolve(msg);
    }
    this.streamResolvers = [];
    for (const resolve of this.controlResponseWaiters.values()) {
      resolve({ subtype: "error", error: "session closed" });
    }
    this.controlResponseWaiters.clear();
  }
  getMessageRunId(msg) {
    switch (msg.type) {
      case "assistant":
      case "tool_call":
      case "tool_result":
      case "reasoning":
      case "error":
      case "retry":
        return msg.runId;
      default:
        return;
    }
  }
  updateCompletedRunIds(resultRunIds, streamedRunIds) {
    const nextRunIds = new Set;
    if (Array.isArray(resultRunIds)) {
      for (const runId of resultRunIds) {
        if (runId) {
          nextRunIds.add(runId);
        }
      }
    }
    for (const runId of streamedRunIds) {
      if (runId) {
        nextRunIds.add(runId);
      }
    }
    this.lastCompletedRunIds = nextRunIds;
  }
  async registerExternalTools() {
    const toolDefs = Array.from(this.externalTools.values()).map((tool) => ({
      name: tool.name,
      label: tool.label,
      description: tool.description,
      parameters: this.schemaToJsonSchema(tool.parameters)
    }));
    sessionLog("registerTools", `registering ${toolDefs.length} external tools: ${toolDefs.map((t) => t.name).join(", ")}`);
    await this.transport.write({
      type: "control_request",
      request_id: `register_tools_${Date.now()}`,
      request: {
        subtype: "register_external_tools",
        tools: toolDefs
      }
    });
  }
  schemaToJsonSchema(schema) {
    if (schema && typeof schema === "object") {
      const s = schema;
      return {
        type: s.type,
        properties: s.properties,
        required: s.required,
        additionalProperties: s.additionalProperties,
        description: s.description
      };
    }
    return { type: "object" };
  }
  async handleExecuteExternalTool(requestId, req) {
    const tool = this.externalTools.get(req.tool_name);
    if (!tool) {
      sessionLog("executeExternalTool", `ERROR: unknown tool ${req.tool_name}`);
      await this.transport.write({
        type: "control_response",
        response: {
          subtype: "external_tool_result",
          request_id: requestId,
          tool_call_id: req.tool_call_id,
          content: [{ type: "text", text: `Unknown external tool: ${req.tool_name}` }],
          is_error: true
        }
      });
      return;
    }
    try {
      sessionLog("executeExternalTool", `executing ${req.tool_name} (call_id=${req.tool_call_id})`);
      const result = await tool.execute(req.tool_call_id, req.input);
      await this.transport.write({
        type: "control_response",
        response: {
          subtype: "external_tool_result",
          request_id: requestId,
          tool_call_id: req.tool_call_id,
          content: result.content,
          is_error: false
        }
      });
      sessionLog("executeExternalTool", `${req.tool_name} completed successfully`);
    } catch (err) {
      const errorMessage = err instanceof Error ? err.message : String(err);
      sessionLog("executeExternalTool", `${req.tool_name} failed: ${errorMessage}`);
      await this.transport.write({
        type: "control_response",
        response: {
          subtype: "external_tool_result",
          request_id: requestId,
          tool_call_id: req.tool_call_id,
          content: [{ type: "text", text: `Tool execution error: ${errorMessage}` }],
          is_error: true
        }
      });
    }
  }
  async handleCanUseTool(requestId, req) {
    let response;
    const toolName = req.tool_name;
    const hasCallback = typeof this.options.canUseTool === "function";
    const toolNeedsRuntimeUserInput = requiresRuntimeUserInput(toolName);
    const autoAllowWithoutCallback = isHeadlessAutoAllowTool(toolName);
    sessionLog("canUseTool", `tool=${toolName} mode=${this.options.permissionMode || "default"} requestId=${requestId}`);
    if (toolNeedsRuntimeUserInput && !hasCallback) {
      response = {
        behavior: "deny",
        message: "No canUseTool callback registered",
        interrupt: false
      };
    } else if (this.options.permissionMode === "bypassPermissions" && !toolNeedsRuntimeUserInput) {
      sessionLog("canUseTool", `AUTO-ALLOW ${toolName} (bypassPermissions)`);
      response = {
        behavior: "allow",
        updatedInput: null,
        updatedPermissions: []
      };
    } else if (hasCallback) {
      try {
        const result = await this.options.canUseTool(toolName, req.input);
        if (result.behavior === "allow") {
          response = {
            behavior: "allow",
            updatedInput: result.updatedInput ?? null,
            updatedPermissions: []
          };
        } else {
          response = {
            behavior: "deny",
            message: result.message ?? "Denied by canUseTool callback",
            interrupt: false
          };
        }
      } catch (err) {
        response = {
          behavior: "deny",
          message: err instanceof Error ? err.message : "Callback error",
          interrupt: false
        };
      }
    } else if (autoAllowWithoutCallback) {
      sessionLog("canUseTool", `AUTO-ALLOW ${toolName} (default behavior)`);
      response = {
        behavior: "allow",
        updatedInput: null,
        updatedPermissions: []
      };
    } else {
      response = {
        behavior: "deny",
        message: "No canUseTool callback registered",
        interrupt: false
      };
    }
    const responseBehavior = "behavior" in response ? response.behavior : "unknown";
    sessionLog("canUseTool", `responding: requestId=${requestId} behavior=${responseBehavior}`);
    await this.transport.write({
      type: "control_response",
      response: {
        subtype: "success",
        request_id: requestId,
        response
      }
    });
    sessionLog("canUseTool", `response sent for ${toolName}`);
  }
  async abort() {
    sessionLog("abort", `aborting session (agent=${this._agentId})`);
    await this.transport.write({
      type: "control_request",
      request_id: `interrupt-${Date.now()}`,
      request: { subtype: "interrupt" }
    });
  }
  async listMessages(options = {}) {
    if (!this.initialized) {
      throw new Error("Session must be initialized before calling listMessages()");
    }
    const requestId = `list-${Date.now()}-${Math.random().toString(36).slice(2, 7)}`;
    const responsePromise = new Promise((resolve) => {
      this.controlResponseWaiters.set(requestId, resolve);
    });
    await this.transport.write({
      type: "control_request",
      request_id: requestId,
      request: {
        subtype: "list_messages",
        ...options.conversationId ? { conversation_id: options.conversationId } : {},
        ...options.before ? { before: options.before } : {},
        ...options.after ? { after: options.after } : {},
        ...options.order ? { order: options.order } : {},
        ...options.limit !== undefined ? { limit: options.limit } : {}
      }
    });
    const resp = await responsePromise;
    if (!resp) {
      throw new Error("Session closed before listMessages response arrived");
    }
    if (resp.subtype === "error") {
      throw new Error(resp.error ?? "listMessages failed");
    }
    const payload = resp.response;
    return {
      messages: payload?.messages ?? [],
      nextBefore: payload?.next_before ?? null,
      hasMore: payload?.has_more ?? false
    };
  }
  async bootstrapState(options = {}) {
    if (!this.initialized) {
      throw new Error("Session must be initialized before calling bootstrapState()");
    }
    const requestId = `bootstrap-${Date.now()}-${Math.random().toString(36).slice(2, 7)}`;
    const responsePromise = new Promise((resolve) => {
      this.controlResponseWaiters.set(requestId, resolve);
    });
    await this.transport.write({
      type: "control_request",
      request_id: requestId,
      request: {
        subtype: "bootstrap_session_state",
        ...options.limit !== undefined ? { limit: options.limit } : {},
        ...options.order ? { order: options.order } : {}
      }
    });
    const resp = await responsePromise;
    if (!resp) {
      throw new Error("Session closed before bootstrapState response arrived");
    }
    if (resp.subtype === "error") {
      throw new Error(resp.error ?? "bootstrapState failed");
    }
    const payload = resp.response;
    return {
      agentId: payload?.agent_id ?? this._agentId ?? "",
      conversationId: payload?.conversation_id ?? this._conversationId ?? "",
      model: payload?.model,
      tools: payload?.tools ?? [],
      memfsEnabled: payload?.memfs_enabled ?? false,
      messages: payload?.messages ?? [],
      nextBefore: payload?.next_before ?? null,
      hasMore: payload?.has_more ?? false,
      hasPendingApproval: payload?.has_pending_approval ?? false,
      timings: payload?.timings
    };
  }
  close() {
    sessionLog("close", `closing session (agent=${this._agentId}, conversation=${this._conversationId})`);
    this.transport.close();
    this.pumpClosed = true;
    this.resolveAllStreamWaiters(null);
  }
  get agentId() {
    return this._agentId;
  }
  get sessionId() {
    return this._sessionId;
  }
  get conversationId() {
    return this._conversationId;
  }
  async[Symbol.asyncDispose]() {
    this.close();
  }
  transformMessage(wireMsg) {
    if (wireMsg.type === "system" && "subtype" in wireMsg && wireMsg.subtype === "init") {
      const msg = wireMsg;
      return {
        type: "init",
        agentId: msg.agent_id,
        sessionId: msg.session_id,
        conversationId: msg.conversation_id,
        model: msg.model,
        tools: msg.tools,
        memfsEnabled: msg.memfs_enabled,
        skillSources: msg.skill_sources,
        systemInfoReminderEnabled: msg.system_info_reminder_enabled,
        sleeptime: msg.reflection_trigger && msg.reflection_behavior && typeof msg.reflection_step_count === "number" ? {
          trigger: msg.reflection_trigger,
          behavior: msg.reflection_behavior,
          stepCount: msg.reflection_step_count
        } : undefined
      };
    }
    if (wireMsg.type === "message" && "message_type" in wireMsg) {
      const msg = wireMsg;
      const runId = msg.run_id || undefined;
      if (msg.message_type === "assistant_message" && msg.content) {
        return {
          type: "assistant",
          content: msg.content,
          uuid: msg.uuid,
          runId
        };
      }
      if (msg.message_type === "tool_call_message" || msg.message_type === "approval_request_message") {
        const toolCallRaw = msg.tool_calls?.[0] || msg.tool_call;
        if (toolCallRaw) {
          const fnObj = toolCallRaw.function;
          const toolCallId = toolCallRaw.tool_call_id ?? toolCallRaw.id;
          if (!toolCallId) {
            return null;
          }
          const toolName = toolCallRaw.name ?? fnObj?.name ?? "?";
          const toolArgs = toolCallRaw.arguments ?? fnObj?.arguments ?? "";
          let toolInput = {};
          try {
            toolInput = JSON.parse(toolArgs);
          } catch {
            toolInput = { raw: toolArgs };
          }
          return {
            type: "tool_call",
            toolCallId,
            toolName,
            toolInput,
            rawArguments: toolArgs || undefined,
            uuid: msg.uuid,
            runId
          };
        }
      }
      if (msg.message_type === "tool_return_message" && msg.tool_call_id) {
        return {
          type: "tool_result",
          toolCallId: msg.tool_call_id,
          content: msg.tool_return || "",
          isError: msg.status === "error",
          uuid: msg.uuid,
          runId
        };
      }
      if (msg.message_type === "reasoning_message" && msg.reasoning) {
        return {
          type: "reasoning",
          content: msg.reasoning,
          uuid: msg.uuid,
          runId
        };
      }
    }
    if (wireMsg.type === "stream_event") {
      const msg = wireMsg;
      const eventPayload = msg.event ?? {};
      return {
        type: "stream_event",
        event: eventPayload,
        uuid: msg.uuid
      };
    }
    if (wireMsg.type === "result") {
      const msg = wireMsg;
      const runIds = Array.isArray(msg.run_ids) ? msg.run_ids.filter((id) => typeof id === "string") : undefined;
      return {
        type: "result",
        success: msg.subtype === "success",
        result: msg.result,
        error: msg.subtype !== "success" ? msg.subtype : undefined,
        stopReason: msg.stop_reason,
        durationMs: msg.duration_ms,
        totalCostUsd: msg.total_cost_usd,
        conversationId: msg.conversation_id,
        runIds
      };
    }
    if (wireMsg.type === "error") {
      const msg = wireMsg;
      return {
        type: "error",
        message: msg.message,
        stopReason: msg.stop_reason,
        runId: msg.run_id,
        apiError: msg.api_error
      };
    }
    if (wireMsg.type === "retry") {
      const msg = wireMsg;
      return {
        type: "retry",
        reason: msg.reason,
        attempt: msg.attempt,
        maxAttempts: msg.max_attempts,
        delayMs: msg.delay_ms,
        runId: msg.run_id
      };
    }
    return null;
  }
}

// src/validation.ts
var VALID_SKILL_SOURCES = [
  "bundled",
  "global",
  "agent",
  "project"
];
function getBlockLabels(memory) {
  return memory.map((item) => {
    if (typeof item === "string")
      return item;
    if ("label" in item)
      return item.label;
    return null;
  }).filter((label) => label !== null);
}
function validateSystemPromptPreset(preset) {
  const validPresets = [
    "default",
    "letta-claude",
    "letta-codex",
    "letta-gemini",
    "claude",
    "codex",
    "gemini"
  ];
  if (!validPresets.includes(preset)) {
    throw new Error(`Invalid system prompt preset '${preset}'. ` + `Valid presets: ${validPresets.join(", ")}`);
  }
}
function validateSkillSources(sources) {
  if (sources === undefined) {
    return;
  }
  for (const source of sources) {
    if (!VALID_SKILL_SOURCES.includes(source)) {
      throw new Error(`Invalid skill source '${source}'. Valid values: ${VALID_SKILL_SOURCES.join(", ")}`);
    }
  }
}
function validateSleeptimeOptions(sleeptime) {
  if (sleeptime === undefined) {
    return;
  }
  if (sleeptime.trigger !== undefined && !["off", "step-count", "compaction-event"].includes(sleeptime.trigger)) {
    throw new Error(`Invalid sleeptime.trigger '${String(sleeptime.trigger)}'. Valid values: off, step-count, compaction-event`);
  }
  if (sleeptime.behavior !== undefined && !["reminder", "auto-launch"].includes(sleeptime.behavior)) {
    throw new Error(`Invalid sleeptime.behavior '${String(sleeptime.behavior)}'. Valid values: reminder, auto-launch`);
  }
  if (sleeptime.stepCount !== undefined && (!Number.isInteger(sleeptime.stepCount) || sleeptime.stepCount <= 0)) {
    throw new Error("Invalid sleeptime.stepCount. Expected a positive integer.");
  }
}
function validateCreateSessionOptions(options) {
  if (options.systemPrompt !== undefined) {
    validateSystemPromptPreset(options.systemPrompt);
  }
  validateSkillSources(options.skillSources);
  validateSleeptimeOptions(options.sleeptime);
}
function validateCreateAgentOptions(options) {
  if (options.memory !== undefined) {
    const blockLabels = getBlockLabels(options.memory);
    if (options.persona !== undefined && !blockLabels.includes("persona")) {
      throw new Error("Cannot set 'persona' value - block not included in 'memory'. " + "Either add 'persona' to memory array or remove the persona option.");
    }
    if (options.human !== undefined && !blockLabels.includes("human")) {
      throw new Error("Cannot set 'human' value - block not included in 'memory'. " + "Either add 'human' to memory array or remove the human option.");
    }
  }
  if (options.systemPrompt !== undefined && typeof options.systemPrompt === "object") {
    validateSystemPromptPreset(options.systemPrompt.preset);
  } else if (options.systemPrompt !== undefined && typeof options.systemPrompt === "string") {
    const validPresets = [
      "default",
      "letta-claude",
      "letta-codex",
      "letta-gemini",
      "claude",
      "codex",
      "gemini"
    ];
    if (validPresets.includes(options.systemPrompt)) {
      validateSystemPromptPreset(options.systemPrompt);
    }
  }
  validateSkillSources(options.skillSources);
  validateSleeptimeOptions(options.sleeptime);
}
// src/stream-events.ts
function extractTextFromContent(content) {
  if (typeof content === "string") {
    return content;
  }
  if (Array.isArray(content)) {
    const pieces = [];
    for (const part of content) {
      if (typeof part === "string") {
        pieces.push(part);
        continue;
      }
      if (!part || typeof part !== "object") {
        continue;
      }
      const rec = part;
      if (typeof rec.text === "string") {
        pieces.push(rec.text);
      }
    }
    const joined = pieces.join("");
    return joined.length > 0 ? joined : null;
  }
  if (content && typeof content === "object") {
    const rec = content;
    if (typeof rec.text === "string") {
      return rec.text;
    }
  }
  return null;
}
function extractStreamTextDelta(event) {
  if (!event || typeof event !== "object") {
    return null;
  }
  const rec = event;
  const maybeDelta = rec.delta;
  if (maybeDelta && typeof maybeDelta === "object") {
    const delta = maybeDelta;
    if (typeof delta.reasoning === "string" && delta.reasoning.length > 0) {
      return { kind: "reasoning", text: delta.reasoning };
    }
    if (typeof delta.text === "string" && delta.text.length > 0) {
      return { kind: "assistant", text: delta.text };
    }
  }
  const messageType = rec.message_type;
  if (messageType === "reasoning_message") {
    const reasoningText = typeof rec.reasoning === "string" ? rec.reasoning : extractTextFromContent(rec.content);
    if (reasoningText && reasoningText.length > 0) {
      return { kind: "reasoning", text: reasoningText };
    }
  }
  if (messageType === "assistant_message") {
    const assistantText = extractTextFromContent(rec.content);
    if (assistantText && assistantText.length > 0) {
      return { kind: "assistant", text: assistantText };
    }
  }
  return null;
}
// src/tool-helpers.ts
function jsonResult(payload) {
  return {
    content: [
      {
        type: "text",
        text: JSON.stringify(payload, null, 2)
      }
    ],
    details: payload
  };
}
function readStringParam(params, key, options = {}) {
  const { required = false, trim = true, label = key, allowEmpty = false } = options;
  const raw = params[key];
  if (typeof raw !== "string") {
    if (required)
      throw new Error(`${label} required`);
    return;
  }
  const value = trim ? raw.trim() : raw;
  if (!value && !allowEmpty) {
    if (required)
      throw new Error(`${label} required`);
    return;
  }
  return value;
}
function readNumberParam(params, key, options = {}) {
  const { required = false, label = key, integer = false } = options;
  const raw = params[key];
  let value;
  if (typeof raw === "number" && Number.isFinite(raw)) {
    value = raw;
  } else if (typeof raw === "string") {
    const trimmed = raw.trim();
    if (trimmed) {
      const parsed = Number.parseFloat(trimmed);
      if (Number.isFinite(parsed))
        value = parsed;
    }
  }
  if (value === undefined) {
    if (required)
      throw new Error(`${label} required`);
    return;
  }
  return integer ? Math.trunc(value) : value;
}
function readBooleanParam(params, key, options = {}) {
  const { required = false, label = key } = options;
  const raw = params[key];
  if (typeof raw === "boolean") {
    return raw;
  }
  if (typeof raw === "string") {
    const lower = raw.toLowerCase().trim();
    if (lower === "true" || lower === "1" || lower === "yes")
      return true;
    if (lower === "false" || lower === "0" || lower === "no")
      return false;
  }
  if (required)
    throw new Error(`${label} required`);
  return;
}
function readStringArrayParam(params, key, options = {}) {
  const { required = false, label = key } = options;
  const raw = params[key];
  if (Array.isArray(raw)) {
    const values = raw.filter((entry) => typeof entry === "string").map((entry) => entry.trim()).filter(Boolean);
    if (values.length === 0) {
      if (required)
        throw new Error(`${label} required`);
      return;
    }
    return values;
  }
  if (typeof raw === "string") {
    const value = raw.trim();
    if (!value) {
      if (required)
        throw new Error(`${label} required`);
      return;
    }
    return [value];
  }
  if (required)
    throw new Error(`${label} required`);
  return;
}

// src/index.ts
import { readFileSync } from "node:fs";
async function createAgent(options = {}) {
  validateCreateAgentOptions(options);
  const session = new Session({ ...options, createOnly: true });
  const initMsg = await session.initialize();
  session.close();
  return initMsg.agentId;
}
function createSession(agentId, options = {}) {
  validateCreateSessionOptions(options);
  if (agentId) {
    return new Session({ ...options, agentId, newConversation: true });
  } else {
    return new Session({ ...options, newConversation: true });
  }
}
function resumeSession(id, options = {}) {
  validateCreateSessionOptions(options);
  if (id.startsWith("conv-")) {
    return new Session({ ...options, conversationId: id });
  } else {
    return new Session({ ...options, agentId: id, defaultConversation: true });
  }
}
async function prompt(message, agentId) {
  const session = agentId ? createSession(agentId) : createSession();
  try {
    await session.send(message);
    let result = null;
    for await (const msg of session.stream()) {
      if (msg.type === "result") {
        result = msg;
        break;
      }
    }
    if (!result) {
      return {
        type: "result",
        success: false,
        error: "No result received",
        durationMs: 0,
        conversationId: session.conversationId
      };
    }
    return result;
  } finally {
    session.close();
  }
}
async function listMessagesDirect(agentId, options = {}) {
  const session = resumeSession(agentId, {
    permissionMode: "bypassPermissions",
    memfsStartup: "skip",
    skillSources: [],
    systemInfoReminder: false
  });
  await session.initialize();
  try {
    return await session.listMessages(options);
  } finally {
    session.close();
  }
}
function imageFromFile(filePath) {
  const data = readFileSync(filePath).toString("base64");
  const ext = filePath.toLowerCase();
  const media_type = ext.endsWith(".png") ? "image/png" : ext.endsWith(".gif") ? "image/gif" : ext.endsWith(".webp") ? "image/webp" : "image/jpeg";
  return {
    type: "image",
    source: { type: "base64", media_type, data }
  };
}
function imageFromBase64(data, media_type = "image/png") {
  return {
    type: "image",
    source: { type: "base64", media_type, data }
  };
}
async function imageFromURL(url) {
  const response = await fetch(url);
  const buffer = await response.arrayBuffer();
  const data = Buffer.from(buffer).toString("base64");
  const contentType = response.headers.get("content-type");
  let media_type = "image/png";
  if (contentType?.includes("jpeg") || contentType?.includes("jpg") || url.match(/\.jpe?g$/i)) {
    media_type = "image/jpeg";
  } else if (contentType?.includes("gif") || url.endsWith(".gif")) {
    media_type = "image/gif";
  } else if (contentType?.includes("webp") || url.endsWith(".webp")) {
    media_type = "image/webp";
  }
  return {
    type: "image",
    source: { type: "base64", media_type, data }
  };
}
export {
  resumeSession,
  readStringParam,
  readStringArrayParam,
  readNumberParam,
  readBooleanParam,
  prompt,
  listMessagesDirect,
  jsonResult,
  imageFromURL,
  imageFromFile,
  imageFromBase64,
  extractStreamTextDelta,
  createSession,
  createAgent,
  Session
};

//# debugId=6AF7BACA7C42A3D164756E2164756E21
