// @letta-ai/letta-client/resources/blocks/blocks.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Blocks = void 0;
const tslib_1 = require("../../internal/tslib.js");
const resource_1 = require("../../core/resource.js");
const AgentsAPI = tslib_1.__importStar(require("./agents.js"));
const agents_1 = require("./agents.js");
const pagination_1 = require("../../core/pagination.js");
const path_1 = require("../../internal/utils/path.js");
class Blocks extends resource_1.APIResource {
    constructor() {
        super(...arguments);
        this.agents = new AgentsAPI.Agents(this._client);
    }
    /**
     * Create Block
     */
    create(body, options) {
        return this._client.post('/v1/blocks/', { body, ...options });
    }
    /**
     * Retrieve Block
     */
    retrieve(blockID, options) {
        return this._client.get((0, path_1.path) `/v1/blocks/${blockID}`, options);
    }
    /**
     * Update Block
     */
    update(blockID, body, options) {
        return this._client.patch((0, path_1.path) `/v1/blocks/${blockID}`, { body, ...options });
    }
    /**
     * List Blocks
     */
    list(query = {}, options) {
        return this._client.getAPIList('/v1/blocks/', (pagination_1.ArrayPage), { query, ...options });
    }
    /**
     * Delete Block
     */
    delete(blockID, options) {
        return this._client.delete((0, path_1.path) `/v1/blocks/${blockID}`, options);
    }
}
exports.Blocks = Blocks;
Blocks.Agents = agents_1.Agents;
