// @letta-ai/letta-client/resources/steps/steps.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Steps = void 0;
const tslib_1 = require("../../internal/tslib.js");
const resource_1 = require("../../core/resource.js");
const FeedbackAPI = tslib_1.__importStar(require("./feedback.js"));
const feedback_1 = require("./feedback.js");
const StepsMessagesAPI = tslib_1.__importStar(require("./messages.js"));
const messages_1 = require("./messages.js");
const MetricsAPI = tslib_1.__importStar(require("./metrics.js"));
const metrics_1 = require("./metrics.js");
const TraceAPI = tslib_1.__importStar(require("./trace.js"));
const trace_1 = require("./trace.js");
const pagination_1 = require("../../core/pagination.js");
const path_1 = require("../../internal/utils/path.js");
class Steps extends resource_1.APIResource {
    constructor() {
        super(...arguments);
        this.metrics = new MetricsAPI.Metrics(this._client);
        this.trace = new TraceAPI.Trace(this._client);
        this.feedback = new FeedbackAPI.Feedback(this._client);
        this.messages = new StepsMessagesAPI.Messages(this._client);
    }
    /**
     * Get a step by ID.
     */
    retrieve(stepID, options) {
        return this._client.get((0, path_1.path) `/v1/steps/${stepID}`, options);
    }
    /**
     * List steps with optional pagination and date filters.
     */
    list(query = {}, options) {
        return this._client.getAPIList('/v1/steps/', (pagination_1.ArrayPage), { query, ...options });
    }
}
exports.Steps = Steps;
Steps.Metrics = metrics_1.Metrics;
Steps.Trace = trace_1.Trace;
Steps.Feedback = feedback_1.Feedback;
Steps.Messages = messages_1.Messages;
