// @letta-ai/letta-client/resources/folders/files.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Files = void 0;
const resource_1 = require("../../core/resource.js");
const pagination_1 = require("../../core/pagination.js");
const headers_1 = require("../../internal/headers.js");
const uploads_1 = require("../../internal/uploads.js");
const path_1 = require("../../internal/utils/path.js");
class Files extends resource_1.APIResource {
    /**
     * Retrieve a file from a folder by ID.
     */
    retrieve(fileID, params, options) {
        const { folder_id, ...query } = params;
        return this._client.get((0, path_1.path) `/v1/folders/${folder_id}/files/${fileID}`, { query, ...options });
    }
    /**
     * List paginated files associated with a data folder.
     */
    list(folderID, query = {}, options) {
        return this._client.getAPIList((0, path_1.path) `/v1/folders/${folderID}/files`, (pagination_1.ArrayPage), {
            query,
            ...options,
        });
    }
    /**
     * Delete a file from a folder.
     */
    delete(fileID, params, options) {
        const { folder_id } = params;
        return this._client.delete((0, path_1.path) `/v1/folders/${folder_id}/${fileID}`, {
            ...options,
            headers: (0, headers_1.buildHeaders)([{ Accept: '*/*' }, options?.headers]),
        });
    }
    /**
     * Upload a file to a data folder.
     */
    upload(folderID, params, options) {
        const { duplicate_handling, name, ...body } = params;
        return this._client.post((0, path_1.path) `/v1/folders/${folderID}/upload`, (0, uploads_1.multipartFormRequestOptions)({ query: { duplicate_handling, name }, body, ...options }, this._client));
    }
}
exports.Files = Files;
