// @letta-ai/letta-client/resources/folders/folders.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Folders = void 0;
const tslib_1 = require("../../internal/tslib.js");
const resource_1 = require("../../core/resource.js");
const AgentsAPI = tslib_1.__importStar(require("./agents.js"));
const agents_1 = require("./agents.js");
const FilesAPI = tslib_1.__importStar(require("./files.js"));
const files_1 = require("./files.js");
const pagination_1 = require("../../core/pagination.js");
const path_1 = require("../../internal/utils/path.js");
class Folders extends resource_1.APIResource {
    constructor() {
        super(...arguments);
        this.files = new FilesAPI.Files(this._client);
        this.agents = new AgentsAPI.Agents(this._client);
    }
    /**
     * Create a new data folder.
     */
    create(body, options) {
        return this._client.post('/v1/folders/', { body, ...options });
    }
    /**
     * Get a folder by ID
     */
    retrieve(folderID, options) {
        return this._client.get((0, path_1.path) `/v1/folders/${folderID}`, options);
    }
    /**
     * Update the name or documentation of an existing data folder.
     */
    update(folderID, body, options) {
        return this._client.patch((0, path_1.path) `/v1/folders/${folderID}`, { body, ...options });
    }
    /**
     * List all data folders created by a user.
     */
    list(query = {}, options) {
        return this._client.getAPIList('/v1/folders/', (pagination_1.ArrayPage), { query, ...options });
    }
    /**
     * Delete a data folder.
     */
    delete(folderID, options) {
        return this._client.delete((0, path_1.path) `/v1/folders/${folderID}`, options);
    }
}
exports.Folders = Folders;
Folders.Files = files_1.Files;
Folders.Agents = agents_1.Agents;
